### PR TITLE
Add Semantic/* and Relation/* ancestor types

### PR DESCRIPTION
**Pull request by camsaul**
This adds new `Semantic/*` and `Relation/*` ancestor types to our type system. Semantic types such as `:type/Name` derive from `:Semantic/*`, and relation types (`:type/PK` and `:type/FK`) derive from `:Relation/*`. I've added new schemas to validate these accordingly. 

Other small changes: 
- I fixed a few weird things in `types.cljc` (such as duplicate type definitions).
- I removed places where we used `:type/Number` as a semantic type. `:count` aggregations and the like that used to come back with `{:semantic_type :type/Number}` now come back with `:type/Quantity` (an actual semantic type) or nothing. In some cases we were returning `{:semantic_type :type/Number}` for non-numeric expression or `:case` aggregations. I improved the auto-detection logic a little here.
- A few unrelated tweaks so that sync failures when loading test data bubble the exception to the top rather than swallowing/logging it

The eventual goal is to change the prefix of semantic and relation types to `:Semantic/` and `:Relation/` respectively  (see #15956). This PR is a subset of that project and avoids renaming anything.

### CODE REVIEW

File: test/metabase/query_processor/middleware/annotate_test.clj [hunk 1]
```diff
@@ -235,14 +235,13 @@
...
       (is (= {:base_type       :type/Float
-              :semantic_type   :type/Number
               :name            "double-price"
               :display_name    "double-price"
               :expression_name "double-price"
               :field_ref       [:expression "double-price"]}
              (mt/$ids venues
                (#'annotate/col-info-for-field-clause
-                {:expressions {"double-price" [:* $price 2]}}
+                   {:expressions {"double-price" [:* $price 2]}}
```
Comment (dpsutton) [inline test/metabase/query_processor/middleware/annotate_test.clj:244]: this got indented weirdly for no reason

File: test/metabase/query_processor/middleware/annotate_test.clj [hunk 2]
```diff
@@ -413,65 +412,50 @@
...
                                     [:+ [:field (mt/id :venues :price) nil] 1]]]])))
   (testing "Make sure we skip nils when infering case return type"
-    (is (= {:base_type    :type/Float
-            :semantic_type :type/Number}
+    (is (= {:base_type :type/Number}
            (infered-col-type [:case [[[:< [:field (mt/id :venues :price) nil] 10] [:value nil {:base_type :type/Number}]]
                                      [[:> [:field (mt/id :venues :price) nil] 2] 10]]]))))
-  (is (= {:base_type    :type/Float
-          :semantic_type :type/Number}
+  (is (= {:base_type :type/Float}
```
Comment (dpsutton) [inline test/metabase/query_processor/middleware/annotate_test.clj:451]: this looks so much better. why were we ever inferring a semantic type?

**Review by dpsutton — APPROVED** [review]

**Comment by dpsutton** [comment]
Interesting failure from redshift:
> S3ServiceException:The S3 bucket addressed by the query is in a different region from this cluster.

File: test/metabase/query_processor/middleware/annotate_test.clj (hunk 1 above)
Comment (camsaul) [reply test/metabase/query_processor/middleware/annotate_test.clj:244]: oops

**Pull request merged**